### ISSUE TITLE
Using outdated shoulda-matchers version

### ISSUE THREAD

**Issue by mkllnk** [thoughtbot/shoulda#275]
I just noticed that we can update shoulda-matchers after removing this gem. There seem to be some incompatibilities between shoulda-matchers 4 and newer Rails versions. Since there's not much time for maintaining this gem, I'm wondering about these options:

- Remove version restrictions.
- Update Readme and archive.

**Comment by mcmire** [comment]
As infrequently as this gem is updated, I don't think removing the version restrictions would be a good idea, because it goes against expectations around gems and backward compatibility. Specifically, it means that two developers who ran `bundle install` on the same project on different days could get different results. I think it would make more sense to bump `shoulda-matchers` appropriately and then bump the major version.

**Comment by mkllnk** [comment]
> two developers who ran bundle install on the same project on different days could get different results

Oh, all my projects use the Gemfile.lock to ensure that this doesn't happen. I didn't know that there are different setups where this could be a problem. So if you are happy to update, that's great. 

Feel free to ignore my suggestions. The current issue is just that it hasn't been updated and that caused incompatibilities in our code.

**Comment by TastyPi** [comment]
I also ran into this issue as I thought there was a bug in `shoulda-matchers`, but it just turned out we're using `shoulda` and that pulls in an old version. I've replaced `shoulda` with `shoulda-context` and `shoulda-matchers` in our Gemfile to avoid this issue, but either this package should be updated or it should be deprecated with a big warning saying not to use it.

https://github.com/thoughtbot/shoulda-matchers/issues/1529

**Comment by pboling** [comment]
Yeah, this makes no sense.  It used to be the case that bundler advised gem authors to not commit `Gemfile.lock`, but this is no longer the case.  There is no longer a use case I am aware of where it is expected that developers will not be using `Gemfile.lock`.

Our Gemfile:
```
  gem 'shoulda'
  gem 'shoulda-matchers'
```

When running `bundle update`, this is the resulting change:

<img width="259" alt="Screenshot 2023-02-20 at 1 17 41 PM" src="https://user-images.githubusercontent.com/19505/220026246-cd4aa0ed-cc7d-416e-bf8d-eb055f4b1361.png">

It downgrades from `shoulda` v4 to v2 just so it can pull in the latest version of shoulda-matchers v5.

This is worse than having a vague version constraint that would allow bundler to "do the right thing", I.e. get the latest versions of everything it can.

**Comment by mcmire** [comment]
Sorry folks, we've neglected this gem it seems. I'll work on preparing a new release that upgrades this gem to the current version of `shoulda-matchers` and make sure that the versions of Ruby/Rails match by the end of the week.

**Comment by pboling** [comment]
@mcmire Looks like you're not at Thoughtbot anymore.  Do you know if someone else be picking this up, are are you still involved?

**Comment by mcmire** [comment]
Thanks for the ping, everyone, and sorry for the lack of response here. Things have been a bit busy lately. 

I dusted this gem off again and there were some things that were pretty out of date. For one, I couldn't even run the tests anymore because it was relying on a version of Ruby that doesn't compile on recent versions of macOS anymore. Lately we've been making some upgrades to `shoulda-matchers` — dropping support for old Rubies and Rails and adding support for more recent ones. We are likely going to issue a new release of `shoulda-matchers` relatively soon, but in the meantime I've applied those same updates here.

So, I just released 5.0.0.rc1 which includes these updates and, of course, bumps the dependency on `shoulda-matchers` to 5.x. I'll look into seeing how much work we have left on `shoulda-matchers` and hopefully once we release 6.0.0 I can bump the dependency over here again and issue a proper release. In the meantime, please try this out in your project and let me know if you run into any issues.

**Comment by schadenfred** [comment]
> I dusted this gem off again and there were some things that were pretty out of date. For one, I couldn't even run the tests anymore because it was relying on a version of Ruby that doesn't compile on recent versions of macOS anymore. Lately we've been making some upgrades to `shoulda-matchers` — dropping support for old Rubies and Rails and adding support for more recent ones. We are likely going to issue a new release of `shoulda-matchers` relatively soon, but in the meantime I've applied those same updates here.

I'm working on a project that sets gem authors up with a docker image for each version of ruby they'd like to run their gems tests against. Let me know if you think that would be helpful. Otherwise, I appreciate your work!

**Comment by schadenfred** [comment]
> I dusted this gem off again and there were some things that were pretty out of date. For one, I couldn't even run the tests anymore because it was relying on a version of Ruby that doesn't compile on recent versions of macOS anymore. Lately we've been making some upgrades to `shoulda-matchers` — dropping support for old Rubies and Rails and adding support for more recent ones. We are likely going to issue a new release of `shoulda-matchers` relatively soon, but in the meantime I've applied those same updates here.

I'm working on a project that sets gem authors up with a docker image for each version of ruby they'd like to run their gems tests against. Let me know if you think that would be helpful. Otherwise, I appreciate your work!

**Comment by chaadow** [comment]
Hi @mcmire and thanks for releasing `5.0.0.rc1`

Since `6.0.0` has been released, would you mind bumping the dependency? Thanks!

**Comment by mcmire** [comment]
@matsales @vsppedro Making y'all aware of this issue, in case you haven't seen it :)

**Comment by schadenfred** [comment]
> @schadenfred Please do say more about this. I have a project of RuboCop wrappers ([called `rubocop-lts`](https://rubocop-lts.gitlab.io/)) for every version of Ruby going back to Ruby 1.8, and could really use this. Also I maintain a bunch of RSpec plugins which still support Ruby 1.8, since RSpec hasn't dropped it yet.

I'd like to see if I can help. Broad strokes -- The project is called RoRo. You can think of it as a framework for writing configuration backstories. One of those is for gem authors but it's pretty specific to my gems. I haven't had to test any of my gems against ruby 1.8 but I looked around and found a few examples of alpine images and more of the bigger ubuntu based ones, which may not be such a big deal depending on what you'd want to use for ci. I like Circle a lot. Do you have experience with it? 

Is it okay if I reach out to your at the email on your profile to discuss?

**Comment by vpiau** [comment]
> Hi [@mcmire](https://github.com/mcmire) and thanks for releasing `5.0.0.rc1`
> 
> Since `6.0.0` has been released, would you mind bumping the dependency? Thanks!

Hi [@mcmire](https://github.com/mcmire), is there any plan to release new version with `shoulda-matchers 6.0.0` support ?
Thanks

**Comment by pboling** [comment]
@schadenfred [RoRo](https://github.com/handsomefencer/unstoppable) was a bit hard to find, will take discussion there. :)

**Comment by schadenfred** [comment]
> [@schadenfred](https://github.com/schadenfred) [RoRo](https://github.com/handsomefencer/unstoppable) was a bit hard to find, will take discussion there. :)

I put an issue in for a story encompassing what I understand to be your use case: 

https://github.com/handsomefencer/unstoppable/issues/36

We can continue the discussion there.
